### ISSUE TITLE
sometimes|required fails due to Validator method passesOptionalCheck returning true

### ISSUE THREAD

**Issue by MoHangingWithLegends** [laravel/framework#22428]
- Laravel Version: 5.5.17
- PHP Version: 7.0.10
- Database Driver & Version: N/A

### Description:
Rule sometimes|required does not behave as expected for collections.
https://laravel.com/docs/5.5/validation#conditionally-adding-rules

I want to validate the field as required if present. Therefore removing required and nullable did not work.

I believe its because Validator passesOptionalCheck() is returning true while it should be returning false.
This means the required validation is not skipped even though the attributes are not present in the data.

As I am new I do apologies beforehand. If the question is not to standard.

### Steps To Reproduce:

```
METHOD: PUT
Content-Type: application/json

json:
[
    {"id" : "93049", "port" : "5"},
    {"id" : "93050", "password" : "55555", "version" : "5"}
]
```

```
$rules = [
    '*.type' => 'sometimes|required|string'
];

Using lumen:
$this->validate($request, $rules);
```

**Response:**
```
{
    "errors": [
        {
            "title": "Validation Error",
            "detail": "The type field is required",
            "status": 422,
            "source": "0.type"
        },
        {
            "title": "Validation Error",
            "detail": "The type field is required",
            "status": 422,
            "source": "1.type"
        }
.......
}
```

**$request->request**:
```
object(Symfony\Component\HttpFoundation\ParameterBag)[46]
  protected 'parameters' => 
    array (size=2)
      0 => 
        array (size=3)
          'id' => string '93049' (length=5)
          'port' => string '5' (length=1)
          'server_id' => string '4222' (length=4)
      1 => 
        array (size=4)
          'id' => string '93050' (length=5)
          'password' => string '55555' (length=5)
          'version' => string '5' (length=1)
          'server_id' => string '4222' (length=4)
```

**Debugging:**
**modified the method to pass the $rule** for debugging:
```
    protected function passesOptionalCheck($attribute, $rule = null)
    {
        if (! $this->hasRule($attribute, ['Sometimes'])) {
            return true;
        }
        $data = ValidationData::initializeAndGatherData($attribute, $this->data);

        if ($attribute === "0.type" && $rule === 'Required') {
            dd( array_key_exists($attribute, $data)); //returns false (ok)
            dd( in_array($attribute, array_keys($this->data))); // returns true (problem)
            dd( array_keys($this->data)); // gives [0, 1]
            dd( in_array( '0.type', [0,1])); // returns true (problem)
            dd(in_array( '0.type', [0,1], true)); //returns false (problem solved)
        }

        return array_key_exists($attribute, $data)
                    || in_array($attribute, array_keys($this->data));
    }
```

using strict comparison fixes the problem for me:
`
in_array($attribute, array_keys($this->data), true);
`

However the check should be against keys within the array and not the index.

`$data = ValidationData::initializeAndGatherData($attribute, $this->data);`

$data is fine as it returns empty as the field is not present in data but `in_array($attribute, array_keys($this->data))` uses `$this->data`.

using $data instead of $this->data fixes the problem for me:
`
in_array($attribute, array_keys($data));
`

### Extra info:
illuminate/validation: v5.5.17

PHP 7.0.10 (cli) (built: Nov  3 2016 08:06:03) ( NTS )
Copyright (c) 1997-2016 The PHP Group
Zend Engine v3.0.0, Copyright (c) 1998-2016 Zend Technologies
    with Xdebug v2.5.5, Copyright (c) 2002-2017, by Derick Rethans

**Comment by fabpin** [comment]
already have the same error
First of all, Taylor Otwell, thanks for laravel, excellent work, and thanks for sharing this fabulous framework with us.

When sending a post for creating a sponsor, appear than accept the request, but didn't enter the request to the function, so don't create the sponsor but redirect the view with no issues reported on $errors array, 

- Laravel Version: 5.5.33
- PHP Version: 7.1.13
- Database Driver & Version: MySql Version 5.7.21 Driver PHP7.1-msql
- Web service: Nginx 1.10.3
- Dropplet on Digital Ocean, with OS Ubuntu 16.04 of 64bits

Already I have this file on app/Http/Request/SponsorRequest: 
```
<?php

namespace App\Http\Requests;

use Illuminate\Foundation\Http\FormRequest;

class SponsorRequest extends FormRequest
{
    /**
     * Determine if the user is authorized to make this request.
     *
     * @return bool
     */
    public function authorize()
    {
        return true;
    }

    /**
     * Get the validation rules that apply to the request.
     *
     * @return array
     */
    public function rules()
    {
        return [
          'name' => 'required|unique:sponsors,name|string|max:255',
          'description' => 'required|string|max:255',
          'number_of_travels' => 'required|numeric|max:255',
          'name_package' => 'required|unique:travels_hired,name|string|max:255',
          'file_svg' => 'required|image',
        ];
    }
}

```

file on app/Http/Controller/SponsorController(function create):

```
<?php

namespace App\Http\Controllers;

use DB;
use App\Sponsor;
use \Carbon\Carbon;
use App\TravelHired;
use App\PVPeopleRequest;
use App\RequestBikeAngel;
use Illuminate\Http\Request;
use Maatwebsite\Excel\Facades\Excel;
use App\Http\Requests\SponsorRequest;
use App\Http\Requests\PackageRequest;
use Illuminate\Support\Facades\Storage;
use App\Http\Requests\SponsorUpdateRequest;

class SponsorController extends Controller
{
    protected $params;
    /**
     * Display a listing of the resource.
     *
     * @return \Illuminate\Http\Response
     */
    public function index(Request $request)
    {
        $this->params = array();
        $this->params['notification_created_user'] = $request->has('notification_created_user')?$request->input('notification_created_user'):false;
        $this->params['notification_updated_user'] = $request->has('notification_updated_user')?$request->input('notification_updated_user'):false;
        $this->params['notification_package_create'] = $request->has('notification_package_create')?$request->input('notification_package_create'):false;
        $this->params['sponsors'] = Sponsor::all();
        return view('sponsor.index', $this->params);
    }

    /**
     * Show the form for creating a new resource.
     *
     * @return \Illuminate\Http\Response
     */
    public function create(SponsorRequest $request)
    {
        $this->params = array();
        $this->params['sponsor_new'] = new Sponsor;
        $this->params['sponsor_new']->name = $request->input('name');
        $this->params['sponsor_new']->description = $request->input('description');
        $this->params['sponsor_new']->statuses_id = 1;
        $path = public_path().'/sponsors/logo/';
        $file = $request->file("file_svg");
        $fileName = $file->getClientOriginalName();
        $file->move($path, $fileName);
        $this->params['sponsor_new']->logo = '/sponsors/logo/'.$fileName;
        $this->params['sponsor_new']->save();
        $this->params['travel_hired'] = new TravelHired;
        $this->params['travel_hired']->name = $request->input("name_package");
        $this->params['travel_hired']->number_travels_hired = (Integer) $request->input("number_of_travels");
        $this->params['travel_hired']->sponsors_id = $this->params['sponsor_new']->id;
        $this->params['travel_hired']->travels_consumed = 0;
        $this->params['travel_hired']->save();
        $this->params['notification_created_user'] = true;
        $this->params['notification_updated_user'] = false;
        $this->params['notification_package_create'] = false;
        return redirect()->route('sponsors', $this->params);
    }
}
```

**Comment by staudenmeir** [comment]
This has been fixed in 5.6.18 (#24012) and will be fixed in the next release of 5.5 (#25852).

**Comment by driesvints** [comment]
Thanks everyone!
